### PR TITLE
Handle multiple PhysicalStorageBuffer in same struct

### LINKED ISSUE #230: Structure layout reuse does not enumerate structure member count
The following GLSL fails spvReflectCreateShaderModule when calling ParseDescriptorBlockVariableUsage on the second member "b":
```glsl
#version 460 core
#extension GL_EXT_buffer_reference : require
#extension GL_EXT_shader_explicit_arithmetic_types_float32 : require
#extension GL_EXT_shader_explicit_arithmetic_types_int64 : require
#extension GL_EXT_shader_explicit_arithmetic_types_int32 : require

layout(local_size_x = 1, local_size_y = 1, local_size_z = 1) in;

layout(buffer_reference, buffer_reference_align = 4) readonly buffer ReadVecf
{
    float32_t values[];
};

layout(buffer_reference, buffer_reference_align = 4) writeonly buffer WriteVecf
{
    float32_t values[];
};

layout(push_constant, std430) uniform Parameters
{
    ReadVecf a;
    ReadVecf b;
    WriteVecf c;
    uint64_t n;
} params;

void main()
{
    uint32_t idx = gl_GlobalInvocationID.x;
    uint32_t stride = gl_NumWorkGroups.x * gl_WorkGroupSize.x;
    for (; idx < params.n; idx += stride) {
        params.c.values[idx] = params.a.values[idx] + params.b.values[idx];
    }
}
```

The following GLSL succeeds:
```glsl
#version 460 core
#extension GL_EXT_buffer_reference : require
#extension GL_EXT_shader_explicit_arithmetic_types_float32 : require
#extension GL_EXT_shader_explicit_arithmetic_types_int64 : require
#extension GL_EXT_shader_explicit_arithmetic_types_int32 : require

layout(local_size_x = 1, local_size_y = 1, local_size_z = 1) in;

layout(buffer_reference, buffer_reference_align = 4) readonly buffer ReadVecf
{
    float32_t values[];
};

layout(buffer_reference, buffer_reference_align = 4) readonly buffer ReadVecf2
{
    float32_t values[];
};

layout(buffer_reference, buffer_reference_align = 4) writeonly buffer WriteVecf
{
    float32_t values[];
};

layout(push_constant, std430) uniform Parameters
{
    ReadVecf a;
    ReadVecf2 b;
    WriteVecf c;
    uint64_t n;
} params;

void main()
{
    uint32_t idx = gl_GlobalInvocationID.x;
    uint32_t stride = gl_NumWorkGroups.x * gl_WorkGroupSize.x;
    for (; idx < params.n; idx += stride) {
        params.c.values[idx] = params.a.values[idx] + params.b.values[idx];
    }
}
```

I haven't narrowed down exactly why this is the case yet, but I at least have a work around for now.

### LINKED ISSUE #231: buffer_reference through push constant causing Assert/crash in spirv-reflect
I am trying to use spirv-reflect to find out which members are statically referenced in a push_constant containing multiple buffer_references.  This code is generated from slang, but I've stripped it down to create a minimal repro:

```
#version 450
#extension GL_EXT_buffer_reference : require
#extension GL_ARB_enhanced_layouts : require
#extension GL_EXT_nonuniform_qualifier : require
#extension GL_KHR_shader_subgroup_arithmetic : require
layout(row_major) uniform;
layout(row_major) buffer;

struct BDAGlobals_t_0
{
    vec4 g_vTest;
    vec4 g_vTest2;
};

layout(buffer_reference, std430, buffer_reference_align = 16) readonly buffer BufferPointer_BDAGlobals_t_0_1
{
    BDAGlobals_t_0 _data;
};

struct GlobalsBDAPushConstant_t_0
{
    BufferPointer_BDAGlobals_t_0_1  g_GlobalsBDAPerStage_0[6];
};


layout(push_constant)
layout(std140) uniform _S2
{
    BufferPointer_BDAGlobals_t_0_1  g_GlobalsBDAPerStage_0[6];
} g_GlobalsBDAPushConstant_0;


struct PS_OUTPUT_0
{
    vec4 vColor_1;
};

layout(location = 0) out vec4 _S149;

void main()
{


    _S149 = g_GlobalsBDAPushConstant_0.g_GlobalsBDAPerStage_0[0]._data.g_vTest;

    return;
}
```

If you compile this with glslangValidator as follows (or using the attached test.psv):
```
> glslangValidator -V spirv_reflect_bda.glsl -S frag -o test.spv
```

Then try to run it through spirv-reflect in a debug build (I'm on main to 3f468129720eded0cef4077302e491036d099856):
```
> spirv-reflect.exe test.spv
Assertion failed: base_var != 0, file C:\Users\dang\src\SPIRV-Reflect\spirv_reflect.c, line 631
```
@greg-lunarg had made patches to support reflecting statically referenced members in a buffer_reference, but it appears to be breaking down on this use case.

[test.zip](https://github.com/KhronosGroup/SPIRV-Reflect/files/13244940/test.zip)

### PR DESCRIPTION
**Pull request by spencer-lunarg**
closes #230
closes #231 

@muraj and @danginsburg if you want to make sure this works for both your use cases that would be helpful

### CODE REVIEW

File: common/output_stream.cpp [hunk 1]
```diff
@@ -938,6 +938,9 @@ std::string ToStringComponentType(const SpvReflectTypeDescription& type,
   return ss.str();
 }
 
+// TODO 212 - Formatting for buffer refference needs to be improved
+static std::unordered_set<uint32_t> physical_pointer_spirv_id;
```
Comment (spencer-lunarg) [inline common/output_stream.cpp:942]: `-y` works fine, but the normal default formatting still needs improvements

File: spirv_reflect.c [hunk 2]
```diff
@@ -2448,6 +2477,18 @@ static SpvReflectResult ParseDescriptorBlockVariable(
             }
           }
           if (!found_recursion) {
+            uint32_t struct_id = FindType(p_module, p_member_type->id)
+                                     ->struct_type_description->id;
+            p_parser
```
Comment (spencer-lunarg) [inline spirv_reflect.c:2482]: @chaoticbob I'm going to just make a PR to run clang-format with more than 80 lines as this starts to just become unnecessarily hard to read

File: tests/glsl/buffer_handle_1.spv.yaml [hunk 3]
```diff
@@ -320,398 +320,3 @@ all_type_descriptions:
     members:
       - *td13
       - *td16
-  - &td18
```
Comment (spencer-lunarg) [inline tests/glsl/buffer_handle_1.spv.yaml:323]: Looked, these were duplicates (these tests do recursion looping) so this is actually good it was removed

**Comment by muraj** [comment]
Sure, I'll give it a try tonight, thank you very much for such a fast response!

**Comment by muraj** [comment]
Yup, this change is working for me, thanks again!

**Comment by danginsburg** [comment]
I tested this change, it does seem to not assert anymore, but I'm not getting any information about which members of a buffer_reference are statically referenced, I just see this in the output:

`Push constant blocks: 1

    0:
      spirv id : 1649
      name     : g_GlobalsBDAPushConstant_0 (_S2)
          // size = 96, padded size = 96
          struct _S2 {

              // abs offset = 0, rel offset = 0, size = 96, padded size = 96, array stride = 16
              ref struct BufferPointer_BDAGlobals_t_0_1 {
              } g_GlobalsBDAPerStage_0[6];

          } g_GlobalsBDAPushConstant_0;`

My understanding from @greg-lunarg is that it should be reporting which members are statically referenced.

**Comment by danginsburg** [comment]
Hi @spencer-lunarg - I am trying to look at this change some more.  When I run your change on tests\glsl\buffer_pointer.spv I see:

```
D:\Users\Dan\src\SPIRV-Reflect\tests\glsl>..\..\bin\Debug\spirv-reflect.exe -v 2 buffer_pointer.spv
generator       : Khronos Glslang Reference Front End
source lang     : GLSL
source lang ver : 460
source file     : buffer_pointer.glsl
entry point     : main (stage=PS)


  Output variables: 1

    0:
      spirv id  : 26
      location  : 0
      type      : float4
      semantic  :
      name      : Color
      qualifier :


  Push constant blocks: 1

    0:
      spirv id : 14
      name     : push (PushData)
          // size = 16, padded size = 16
          struct PushData {

              // abs offset = 0, rel offset = 0, size = 8, padded size = 16
              ref struct Data {
              } data_ptr;

          } push;
```
In @greg-lunarg origin PR both in the spirv-reflect output and yaml it would report on the referenced variables g0/g1/g2, i.e.:

``` spirv id : 14
      name     : push (PushData)
          // size = 16, padded size = 16
          struct PushData {

              // abs offset = 0, rel offset = 0, size = 8, padded size = 16
              ref struct Data {
                  float g0; // abs offset = 0, rel offset = 0, size = 4, padded
size = 4 UNUSED
                  float g1; // abs offset = 4, rel offset = 4, size = 4, padded
size = 4
                  float g2; // abs offset = 8, rel offset = 8, size = 4, padded
size = 8 UNUSED
              } data_ptr;

          } push;
```

I am trying also to use the API and I am not seeing how I can reflect out which members pointed to by the buffer_ref's in the push_constant are active.  So it seems this change isn't fully working?

**Comment by spencer-lunarg** [comment]
@danginsburg looking into fixing the print-out logic now

> I am not seeing how I can reflect out which members pointed to by the buffer_ref's in the push_constant are active

this is an orthogonal issue of `spirv-reflect` as the `accessed` flag is only in the `DescriptorBinding` and not currently found at a push constant level (that will require adding logic to mark accesses at a more fine granularity)

**Comment by danginsburg** [comment]
Is this ready for me to test again or are you still working on the PR?

**Comment by spencer-lunarg** [comment]
ready to be tested, now convinced I fixed the issues, added another test as well

**Comment by danginsburg** [comment]
From my testing so far, this PR now does seem to correctly fill in the `SPV_REFLECT_VARIABLE_FLAGS_UNUSED` flag for the unused BDA fields.

File: spirv_reflect.c [hunk 4]
```diff
@@ -2286,26 +2311,44 @@ static SpvReflectResult ParseDescriptorBlockVariable(SpvReflectPrvParser* p_pars
...
 
-      if (p_member_type->op == SpvOpTypePointer) {
-        if (p_member_type->storage_class == SpvStorageClassPhysicalStorageBuffer) {
-          // Need to make sure we haven't started an infinite recursive loop
-          for (uint32_t i = 0; i < p_parser->physical_pointer_count; i++) {
-            if (p_member_type->id == p_parser->physical_pointer_check[i]) {
-              found_recursion = true;
-              break;  // still need to fill in p_member_type values
-            }
+      if (p_member_type->storage_class == SpvStorageClassPhysicalStorageBuffer &&
```
Comment (chaoticbob) [inline spirv_reflect.c:2314]: nit: comparison before && should be wrapped in parens

**Review by chaoticbob — APPROVED** [review]
Just one nit

**Pull request merged**